runtime/serde/serde-cbor/common/src/aws/smithy/kotlin/runtime/serde/cbor/CborUtils.kt

```diff
@@ -0,0 +1,223 @@
+        }
+
+        val argumentBytes = ((argumentByteLength - 1) downTo 0).map { index ->
+            (this shr (index * 8) and 0xffu).toByte()
+        }.toByteArray()
+        return argumentBytes
+    }
+
+    val head = when {
+        argument < 24u -> ((major.ordinal shl 5).toULong() or argument).toByte()
+        argument < 0x100u -> ((major.ordinal shl 5) or Minor.ARG_1.value.toInt()).toByte()
+        argument < 0x10000u -> ((major.ordinal shl 5) or Minor.ARG_2.value.toInt()).toByte()
+        argument < 0x100000000u -> ((major.ordinal shl 5) or Minor.ARG_4.value.toInt()).toByte()
+        else -> ((major.ordinal shl 5) or Minor.ARG_8.value.toInt()).toByte()
+    }
+
+    return byteArrayOf(head, *argument.toByteArray())
+}
+
+internal fun decodeArgument(buffer: SdkBufferedSource): ULong {
+    val minor = buffer.readByte().toUByte() and MINOR_BYTE_MASK
+
+    if (minor < Minor.ARG_1.value) {
+        return minor.toULong()
+    }
+
+    val numBytes = when (minor) {
+        Minor.ARG_1.value -> 1L
+        Minor.ARG_2.value -> 2L
+        Minor.ARG_4.value -> 4L
+        Minor.ARG_8.value -> 8L
+        else -> throw DeserializationException("Unsupported minor value $minor, expected one of ${Minor.ARG_1.value}, ${Minor.ARG_2.value}, ${Minor.ARG_4.value}, ${Minor.ARG_8.value}")
+    }
+
+    val bytes = SdkBuffer().use {
+        buffer.read(it, numBytes).also { rc ->
+            if (numBytes != rc) throw DeserializationException("Unexpected end of payload. Expected $numBytes, read $rc.")
+        }
+        it.readByteArray()
+    }
+    return bytes.toULong()
+}
+
+// Convert a ByteArray to ULong by left-shifting each byte appropriately
+internal fun ByteArray.toULong() = foldIndexed(0uL) { i, acc, byte ->
+    acc or (byte.toUByte().toULong() shl ((size - 1 - i) * 8))
+}
+
+// Subtracts one from the given BigInteger
+internal fun BigInteger.minusOne(): BigInteger {
+    val digits = toString().toCharArray()
+    var index = digits.lastIndex
+
+    // Process the digits from right to left
+    while (index >= 0) {
+        if (digits[index] > '0') {
+            digits[index] = digits[index] - 1
+            break
+        } else {
+            digits[index] = '9'
+            index--
+        }
+    }
+
+    // Remove leading zeroes
+    val result = digits.concatToString().trimStart('0')
+
+    return if (result.isEmpty()) BigInteger("0") else BigInteger(result)
+}
+
+// Adds one to the given BigInteger
+internal fun BigInteger.plusOne(): BigInteger {
+    val digits = toString().toCharArray()
+    var index = digits.lastIndex
+
+    // Process the digits from right to left
+    while (index >= 0) {
+        if (digits[index] == '9') {
+            digits[index] = '0'
+            index--
+        } else {
+            digits[index] = digits[index] + 1
+            return BigInteger(digits.concatToString())
+        }
+    }
+
+    // If all digits were '9', prepend '1'
+    return BigInteger("1${digits.concatToString()}")
+}
```

Comment:
**Suggestion**: Dealing with big integers as strings is pretty inefficient when performing arithmetic. I suggest we add actual `plus` and `minus` operators to our KMP `BigInteger` class and delegate them on the JVM to [`BigInteger.add`](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#add-java.math.BigInteger-) and [`BigInteger.subtract`](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#subtract-java.math.BigInteger-).